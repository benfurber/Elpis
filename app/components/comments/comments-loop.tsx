import React, { Component } from "react";
import { FlatList, StyleSheet, View, ScrollView } from "react-native";

import { Post, NavigationType } from "interfaces";
import { labels } from "labels";
import { colours, layout } from "styles";
import { calculateTotalComments } from "utils";

import { Comment } from "./comment";
import { NoContent } from "./no-content";
import { Title } from "components";

interface Props {
  comments: Post["comments"];
  header: object;
  navigation: NavigationType;
  noComments: string;
  onPress: (number) => void;
  postId: Post["id"];
}

class CommentsLoop extends Component<Props> {
  noComments() {
    if (calculateTotalComments(this.props.comments) === 0) {
      return <NoContent text={this.props.noComments} />;
    }
  }

  title() {
    const { comments } = this.props;

    const titleText = `${comments.length} ${labels.topics}`;

    return (
      <View style={styles.commentsHeadingContainer}>
        <Title text={titleText} />
      </View>
    );
  }

  render() {
    const { navigation, postId } = this.props;

    return (
      <View style={styles.container}>
        <ScrollView>
          {this.props.header}
          {this.title()}

          {this.noComments()}

          <FlatList
            contentContainerStyle={{ paddingBottom: layout.spacingXL }}
            data={this.props.comments}
            keyExtractor={({ id }) => id}
            renderItem={({ item }) => (
              <Comment
                item={item}
                navigation={navigation}
                onPress={commentId => this.props.onPress(commentId)}
                postId={postId}
              />
            )}
          />
        </ScrollView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  commentsHeadingContainer: {
    marginHorizontal: layout.spacing,
    marginVertical: layout.spacingL,
  },
  container: {
    flex: 1,
    paddingBottom: layout.spacingL,
    width: "100%",
  },
  titleContainer: {
    backgroundColor: colours.whiteTransparent,
    borderRadius: layout.borderRadius,
    margin: layout.spacing,
    padding: layout.spacing,
  },
});

export { CommentsLoop };
