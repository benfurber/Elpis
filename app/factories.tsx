import { Comment, Community, Reply, User } from "interfaces";

export const community: Community = {
  avatarPath: require("assets/images/empower_two_women_logo.png"),
  id: "1235we",
  name: "Empodere Duas Mulheres",
  posts: [],
};

export const user: User = {
  avatarPath: require("assets/images/profile-pic-annon.png"),
  id: "1235we",
  name: "May",
};

export const post = {
  author: community,
  comments: [],
  content: null,
  id: "10387-314fs-12asdbj",
  imagePath: require("assets/images/image_post_1.jpg"),
  publishedAt: new Date("2000-01-01"),
};

export const preComment: Comment = {
  author: user,
  content:
    "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
  discussionLevel: 0,
  edited: false,
  id: "21097",
  isAuthorCurrentUser: false,
  publishedAt: new Date("2019-01-01"),
  replies: [],
  totalReplies: 0,
};

export const reply: Reply = {
  author: user,
  comment: preComment,
  content: "Nao.",
  edited: false,
  id: "21097",
  isAuthorCurrentUser: false,
  link: null,
  publishedAt: new Date("2019-01-01"),
};

export const comment: Comment = {
  author: user,
  content:
    "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
  discussionLevel: 0,
  edited: false,
  id: "21097",
  isAuthorCurrentUser: false,
  publishedAt: new Date("2019-01-01"),
  replies: [reply],
  title: "Meu pai fez o que ela mandou",
  totalReplies: 1,
};
