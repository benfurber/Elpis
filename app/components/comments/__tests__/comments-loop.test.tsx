import React from "react";
import { Text } from "react-native";
import { shallow } from "enzyme";

import { comment, post } from "factories";

import { CommentsLoop } from "../comments-loop";

let navigation;

describe("Comments<CommentsLoop>", () => {
  describe("when there are comments", () => {
    it("renders correctly", () => {
      const header = <Text>Mock header</Text>;
      const noComments = "There are no comments :(";
      const onPress = number => jest.fn(number);

      const component = shallow(
        <CommentsLoop
          comments={[comment]}
          header={header}
          navigation={navigation}
          noComments={noComments}
          onPress={onPress}
          postId={post.id}
        />,
      );

      expect(component).toMatchSnapshot();
    });
  });

  describe("when there are no comments", () => {
    it("renders correctly", () => {
      const header = <Text>Mock header</Text>;
      const noComments = "There are no comments :(";
      const onPress = number => jest.fn(number);

      const component = shallow(
        <CommentsLoop
          comments={[]}
          header={header}
          navigation={navigation}
          noComments={noComments}
          onPress={onPress}
          postId={post.id}
        />,
      );

      expect(component).toMatchSnapshot();
    });
  });
});
