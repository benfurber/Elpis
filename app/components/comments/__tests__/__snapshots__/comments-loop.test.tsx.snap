// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments<CommentsLoop> when there are comments renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <ScrollViewMock>
    <Text>
      Mock header
    </Text>
    <View
      style={
        Object {
          "marginHorizontal": 10,
          "marginVertical": 20,
        }
      }
    >
      <Title
        text="1 Tópicos"
      />
    </View>
    <FlatList
      contentContainerStyle={
        Object {
          "paddingBottom": 40,
        }
      }
      data={
        Array [
          Object {
            "author": Object {
              "avatarPath": Object {
                "testUri": "../../../app/assets/images/empower_two_women_logo.png",
              },
              "id": "1235we",
              "name": "Empodere Duas Mulheres",
            },
            "content": "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
            "createdAt": 2019-01-01T00:00:00.000Z,
            "id": "21097",
            "replies": Array [],
            "totalReplies": 0,
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </ScrollViewMock>
</View>
`;

exports[`Comments<CommentsLoop> when there are no comments renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <ScrollViewMock>
    <Text>
      Mock header
    </Text>
    <View
      style={
        Object {
          "marginHorizontal": 10,
          "marginVertical": 20,
        }
      }
    >
      <Title
        text="0 Tópicos"
      />
    </View>
    <NoContent
      text="There are no comments :("
    />
    <FlatList
      contentContainerStyle={
        Object {
          "paddingBottom": 40,
        }
      }
      data={Array []}
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </ScrollViewMock>
</View>
`;
